python/ql/src/experimental/Security-new-dataflow/CWE-022/PathInjection.ql

```diff
@@ -0,0 +1,112 @@
+/**
+ * The query detects the case where a path is not both normalized and _afterwards_ checked.
+ *
+ * It does so by dividing the problematic situation into two cases:
+ *  1. The path is never normalized.
+ *     This is easily detected by using normalization as a sanitizer.
+ *
+ *  2. The path is normalized at least once, but never checked afterwards.
+ *     This is detected by finding the first normalization and then ensure that
+ *     no checks happen after. Since we start from the first normalization,
+ *     we know that the absence of checks means that no normalization has a
+ *     chek after it. (No checks after a second normalization would be ok if
+ *     there was a check between the first and the second.)
```

Comment:
I should perhaps talk explicitly about file paths and taint paths rather than trying to avoid the latter concept and still express temporal logic. Also, the word "earliest" might be better than the word "first". I will see if I can reword it.